h5py/_npystrings.pyx

```diff
@@ -0,0 +1,262 @@
+    res = NpyString_load(info[0].allocator, <npy_packed_static_string*>elem, unpacked)
+    info[0].i += 1
+    # res == -1 if unpacking the string fails, 1 if packed_string is the null string,
+    # and 0 otherwise.
+    return -1 if res == -1 else 0
+
+
+def npystrings_unpack(hid_t space, size_t contig, size_t noncontig, size_t descr,
+                      size_t npoints):
+    """Convert a NpyString array (NumPy's native variable-width strings dtype)
+    to a zero-terminated char**, which is the in-memory representation for a
+    HDF5 variable-width string dataset.
+
+    Parameters
+    ----------
+    space : hid_t
+        HDF5 dataspace ID for the input non-contiguous buffer
+        containing packed NpyStrings
+    contig : void * (actually a char **)
+        Output buffer for the contiguous zero-terminated char **
+    noncontig : void *
+        Input buffer for the non-contiguous packed NpyStrings
+    descr : PyArray_Descr * (actually a PyArray_StringDTypeObject *)
+        NumPy dtype descriptor for the NpyStrings
+    npoints : size_t
+        Number of points to copy
+
+    Returns
+    -------
+    char *
+        A temporary buffer containing the zero-terminated strings.
+        This buffer must be freed after writing to disk.
+
+        Note that this always the same as contig[0] when this
+        function returns, but H5Tconvert may change the contents
+        of contig[0] in place later on.
+
+    Notes
+    -----
+    This function defines a pure-python API interface to _proxy.pyx.
+    This is necessary to allow this module to be conditionally imported,
+    allowing NumPy 1.x to continue working everywhere else.
+    """
+    return <size_t>_npystrings_unpack(space, <void *>contig, <void *>noncontig,
+                                      <PyArray_Descr *>descr, npoints)
+
+
+cdef char * _npystrings_unpack(hid_t space, void *contig, void *noncontig,
+                               PyArray_Descr *descr, size_t npoints):
+    """Cython API interface of npystrings_unpack"""
+    cdef npystrings_unpack_t info
+    cdef size_t total_size
+    cdef size_t cur_size
+    cdef char *zero_terminated = NULL
+    cdef char *zero_terminated_cur = NULL
+    info.unpacked = NULL
+    info.i = 0
+
+    info.allocator = NpyString_acquire_allocator(<PyArray_StringDTypeObject *>descr)
+    if info.allocator is NULL:
+        raise RuntimeError("Failed to acquire string allocator")
+
+    tid = H5Tcreate(H5T_OPAQUE, SIZEOF_NPY_PACKED_STATIC_STRING)
+    try:
+        # Multiple steps needed:
+        # 1. Read npy_packed_static_string[] from NumPy and unpack
+        #    to npy_static_string[]; which is
+        #    {const char* buf, size_t size}[] - NOT zero-terminated
+        info.unpacked = <npy_static_string*>malloc(
+            npoints * sizeof(npy_static_string)
+        )
+        if info.unpacked is NULL:
+            raise MemoryError("Failed to allocate memory")
```

Comment:
We have an `emalloc()` wrapper in `utils.pyx` that calls includes the error handling, if you want to save a few lines.